### PR TITLE
feat(php8 attributes): building on lexer PR

### PR DESCRIPTION
**Pull request by cseufert**
Builds on #724 lexer functionality. 

PHP 8 Attribute Features:
- [x] Class
- [x] Methods
- [x] Method Params
- [x] Class Variables
- [x] Inline Functions fn()..
- [x] Functions
- [x] Anonymous Classes
- [x] Class Constants
- [x] Interfaces

### CODE REVIEW

**Comment by cseufert** [comment]
Hey @ichiriac or @czosel can you guys take a look, I think it can be at least merged to the PHP 8 branch, and I think willbring the Parser up to being able to parse PHP 8.0

**Comment by czosel** [comment]
@cseufert CI is failing because the code seems to be formatted with the current prettier version, while we're still on `v2.0.5` in this repo. I'd prefer to wait with updating prettier until the `php8` branch has landed. Can you run `npm run fix`, which should revert the formatting changes? I just tried to do it myself, but I don't have push permissions on your fork :wink:

**Comment by czosel** [comment]
Support for constants in anonymous classes still seems to be missing:
```php
$object = new #[ObjectAttribute] class () { /* … */ };
```
fails with
```
SyntaxError: Parse Error : syntax error, unexpected '#[' (T_ATTRIBUTE) on line 2
```

File: src/parser/statement.js [hunk 1]
```diff
@@ -150,9 +154,17 @@ module.exports = {
...
+    if (this.token === this.tok.T_ATTRIBUTE) {
+      attrs = this.read_attr_list();
+    }
     switch (this.token) {
-      case this.tok.T_FUNCTION:
-        return this.read_function(false, false);
+      case this.tok.T_FUNCTION: {
+        const result = this.read_function(false, false);
+        result.attrGroups = attrs;
+        attrs = [];
```
Comment (czosel) [inline src/parser/statement.js:165]: Is the reset of `attrs` really necessary?

File: src/lexer/attribute.js [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+        }
+    }
+    if (this.is_LABEL_START()) {
+      this.consume_LABEL();
+      return this.tok.T_STRING;
+    } else if (this.is_NUM()) {
+      return this.consume_NUM();
+    }
+
+    throw new Error("what now? " + JSON.stringify(ch));
```
Comment (czosel) [inline src/lexer/attribute.js:57]: The error message could probably be improved. In `scripting.js`, we have for example:
```php
    throw new Error(
      'Bad terminal sequence "' +
        ch +
        '" at line ' +
        this.yylineno +
        " (offset " +
        this.offset +
        ")"
    );
```

File: src/lexer/attribute.js [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+    let listDepth = 0;
+    let ch = this.input();
+    if (this.is_WHITESPACE()) {
+      do {
+        ch = this.input();
+      } while (this.is_WHITESPACE());
+      this.unput(1);
+      return null;
+    }
+    // process.stderr.write("Token:" + JSON.stringify(ch) + "\n");
```
Comment (czosel) [inline src/lexer/attribute.js:19]: Not needed

**Comment by czosel** [comment]
@cseufert all in all this looks great already! :+1: If you'd prefer to add anonymous class support in a separate PR, let me know. The only thing this still needs to be merged in the `php8` is the formatting change and the cleanup of the few nitpicks I added above.

File: src/parser/statement.js [hunk 4]
```diff
@@ -150,9 +154,17 @@ module.exports = {
    * ```
    */
   read_inner_statement: function () {
+    let attrs = [];
+    if (this.token === this.tok.T_ATTRIBUTE) {
+      attrs = this.read_attr_list();
```
Comment (czosel) [inline src/parser/statement.js:159]: This line doesn't seem to be covered by the tests according to #749 (see https://github.com/glayzzle/php-parser/pull/749). Does that mean attributes in inner statements are already handled somewhere else?

**Comment by cseufert** [comment]
> @cseufert CI is failing because the code seems to be formatted with the current prettier version, while we're still on `v2.0.5` in this repo. I'd prefer to wait with updating prettier until the `php8` branch has landed. Can you run `npm run fix`, which should revert the formatting changes? I just tried to do it myself, but I don't have push permissions on your fork

I am strugling with this, yarn install did install 2.3.5, however now npm is installing it and looking at the lock file, its 2.2.3 as well.
https://github.com/glayzzle/php-parser/blob/14718704b761c462dc980a05c9eed636f5391ede/package-lock.json#L3412

**Comment by cseufert** [comment]
@czosel I think we need to merge the master branch into the PHP8 branch, as there has been a bunch of stuff done on the main master branch that this branch has, but the upstream PHP8 branch is missing.

**Comment by czosel** [comment]
You're right, it seems that prettier has been updated on `master` in the meantime:
```
    "prettier": {
      "version": "2.2.1",
      "resolved": "https://registry.npmjs.org/prettier/-/prettier-2.2.1.tgz",
      "integrity": "sha512-PqyhM2yCjg/oKkFPtTGUojv7gnZAoG80ttl45O6x2Ug/rMJw4wcc9k6aaf2hibP7BGVCCM33gZoGjyvt9mm16Q==",
      "dev": true
    },
```
I'd prefer to keep formatting changes out of this PR though anyway - otherwise this PR will show up in `git blame` all over the place :wink: 

On this branch, the lockfile specifies
```
    "prettier": {
      "version": "2.0.5",
      "resolved": "https://registry.npmjs.org/prettier/-/prettier-2.0.5.tgz",
      "integrity": "sha512-7PtVymN48hGcO4fGjybyBSIWDsLU4H4XlvOHfq91pz9kkGlonzwTfYkaIEwiRg/dAJF9YlbsduBAgtYLi+8cFg==",
      "dev": true
    },
```
and for me installation works as expected (using `npm` instad of `yarn`):
```sh
$ npm install
$ npm ls prettier
php-parser@3.0.2 /home/christian/dev/php-parser
└── prettier@2.0.5 
```

**Comment by czosel** [comment]
> @czosel I think we need to merge the master branch into the PHP8 branch, as there has been a bunch of stuff done on the main master branch that this branch has, but the upstream PHP8 branch is missing.

Agreed - but I'd say let's do that in a separate PR after merging this, so that we get a clean history. Ok?

File: src/lexer/attribute.js (hunk 3 above)
Comment (cseufert) [reply src/lexer/attribute.js]: For reasons that I dont understand, if I dont unput and return null, then all the Attribute names include leading spaces & sometimes newlines, eg:
```diff
              Attribute {
                "args": Array [],
                "kind": "attribute",
-               "name": "Two",
+               "name": "
+       Two",
              },
```

File: src/lexer/attribute.js (hunk 2 above)
Comment (cseufert) [reply src/lexer/attribute.js]: Has been updated as suggested.

File: src/parser/statement.js (hunk 4 above)
Comment (cseufert) [reply src/parser/statement.js:159]: It seems you can declare class methods inside inner_statements, im just not sure how to write a test to test such a case.

File: src/parser/statement.js (hunk 1 above)
Comment (cseufert) [reply src/parser/statement.js]: I believe I have removed all cases where it is not strictly necessary.

**Comment by cseufert** [comment]
@czosel I have added a few more places that you can put attributes(interfaces, anon classes, class constants), I think I have them all now. I have cleaned up the prettier issues.

**Comment by czosel** [comment]
@cseufert Awesome! I'll try to take another look tonight (CET).

**Review by czosel — COMMENTED** [review]
This looks great :+1: 
Just a few minor comments, then we can merge!

File: src/parser/expr.js [hunk 5]
```diff
@@ -691,7 +691,7 @@ module.exports = {
     const list = [];
     do {
       this.expect(this.tok.T_ATTRIBUTE);
-      const attrGr = this.node("attrgroup")();
+      const attrGr = this.node("attrgroup")([]);
```
[the remark this answers is not in the document]
Comment (cseufert) [reply src/parser/expr.js]: Yes, but due to the way the comments are applied with partial application, it appears you cant skip arguments, or the docs stuff gets sent as the attrib parameter, and the location gets put in the docs.

File: src/parser/expr.js [hunk 6]
```diff
@@ -684,19 +684,19 @@ module.exports = {
   },
   read_attr_list() {
     const list = [];
-    do {
-      this.expect(this.tok.T_ATTRIBUTE);
-      const attrGr = this.node("attrgroup")([]);
-      this.next();
-      attrGr.attrs.push(this.read_attribute());
-      while (this.token === ",") {
+    if (this.token === this.tok.T_ATTRIBUTE)
```
[the remark this answers is not in the document]
Comment (cseufert) [reply src/parser/expr.js]: Can we use eslint to enforce this, see #757 

I have fixed it in this case.

**Comment by czosel** [comment]
Nothing from my side! Sorry about the delay.

**Pull request merged**